### ISSUE TITLE
Heap buffer overflow in OP_SEND

### ISSUE THREAD

**Issue by clayton-shopify** [mruby/mruby#3703]
The following input demonstrates a crash:
```ruby
def e
  j{return}
  rescue e 
  ensure
  z = x
end

def q
  e
end

def j
  yield
end
q
```
ASAN report:
```
==2972==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x63000002f408 at pc 0x000105d7862e bp 0x7fff59fe9ef0 sp 0x7fff59fe9ee8
WRITE of size 4 at 0x63000002f408 thread T0
    #0 0x105d7862d in mrb_vm_exec (mruby:x86_64+0x10016c62d)
    #1 0x105d6f234 in mrb_vm_run (mruby:x86_64+0x100163234)
    #2 0x105da3d5f in mrb_top_run (mruby:x86_64+0x100197d5f)
    #3 0x105e7638d in mrb_load_exec (mruby:x86_64+0x10026a38d)
    #4 0x105e771a5 in mrb_load_file_cxt (mruby:x86_64+0x10026b1a5)
    #5 0x105c0de53 in main mruby.c:227
    #6 0x7fffe5638234 in start (libdyld.dylib:x86_64+0x5234)

0x63000002f408 is located 8 bytes to the right of 61440-byte region [0x630000020400,0x63000002f400)
allocated by thread T0 here:
    #0 0x106020520 in wrap_realloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56520)
    #1 0x105d02b45 in mrb_default_allocf (mruby:x86_64+0x1000f6b45)
    #2 0x105c838b8 in mrb_realloc_simple gc.c:203
    #3 0x105c8400e in mrb_realloc gc.c:217
    #4 0x105da47ef in stack_extend_alloc (mruby:x86_64+0x1001987ef)
    #5 0x105d6638f in stack_extend (mruby:x86_64+0x10015a38f)
    #6 0x105d7b6d0 in mrb_vm_exec (mruby:x86_64+0x10016f6d0)
    #7 0x105d6f234 in mrb_vm_run (mruby:x86_64+0x100163234)
    #8 0x105da3d5f in mrb_top_run (mruby:x86_64+0x100197d5f)
    #9 0x105e7638d in mrb_load_exec (mruby:x86_64+0x10026a38d)
    #10 0x105e771a5 in mrb_load_file_cxt (mruby:x86_64+0x10026b1a5)
    #11 0x105c0de53 in main mruby.c:227
    #12 0x7fffe5638234 in start (libdyld.dylib:x86_64+0x5234)

SUMMARY: AddressSanitizer: heap-buffer-overflow (mruby:x86_64+0x10016c62d) in mrb_vm_exec
Shadow bytes around the buggy address:
  0x1c6000005e30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c6000005e40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c6000005e50: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c6000005e60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c6000005e70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x1c6000005e80: fa[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c6000005e90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c6000005ea0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c6000005eb0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c6000005ec0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c6000005ed0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==2972==ABORTING
Abort trap: 6
```
This issue was reported by https://hackerone.com/ahihi

**Comment by matz** [comment]
The heap overflow was fixed by 988e2ce
But the behavior still differs from CRuby. I will investigate.

**Comment by matz** [comment]
It turned out that the difference was caused by a CRuby bug.
